databuilder/file_formats/arrow.py

```diff
@@ -0,0 +1,87 @@
+import datetime
+from itertools import islice
+
+import pyarrow
+
+from databuilder.file_formats.validation import ValidationError, validate_headers
+
+PYARROW_TYPE_MAP = {
+    bool: pyarrow.bool_,
+    # If we added some kind of range specification as part of ColumnSpec we could use
+    # that to select smaller numeric types where appropriate
+    int: pyarrow.int64,
+    float: pyarrow.float64,
+    str: pyarrow.string,
+    datetime.date: pyarrow.date32,
+}
+
+# Reading around bits of old blog posts suggests that we want batches in roughly the
+# single digit megabyte range. Of course, translating this into a number of rows depends
+# on the width of the rows, which isn't constant given we use variable width types like
+# `string`. Plus we're using compression so the on disk size will vary in any case.
+# Given 30 columns, each of an average of 32 bits wide, then 64,000 rows takes about
+# 7.7MB, which seems roughly in the right ballpark.
+ROWS_PER_BATCH = 64000
+
+
+def write_dataset_arrow(filename, results, column_specs):
+    schema = schema_from_column_specs(column_specs)
+    options = pyarrow.ipc.IpcWriteOptions(compression="zstd", use_threads=True)
+
+    with pyarrow.OSFile(str(filename), "wb") as sink:
+        with pyarrow.ipc.new_file(sink, schema, options=options) as writer:
+            for results_batch in batch_and_transpose(results, ROWS_PER_BATCH):
+                record_batch = pyarrow.record_batch(results_batch, schema=schema)
+                writer.write(record_batch)
+
+
+def schema_from_column_specs(column_specs):
+    return pyarrow.schema(
+        pyarrow.field(
+            name,
+            PYARROW_TYPE_MAP[spec.type](),
+            nullable=spec.nullable,
+        )
+        for name, spec in column_specs.items()
+    )
+
+
+def batch_and_transpose(iterable, batch_size):
+    """
+    Takes an iterable over rows and returns an iterator over batches of columns e.g.
+
+    >>> results = batch_and_transpose(
+    ...   [(1, "a"), (2, "b"), (3, "c"), (4, "d")],
+    ...   batch_size=3,
+    ... )
+    >>> list(results)
+    [[(1, 2, 3), ('a', 'b', 'c')], [(4,), ('d',)]]
+
+    This is the structure required by Arrow, which is a columar format.
+    """
+    iterator = iter(iterable)
+
+    def next_transposed_batch():
+        row_batch = islice(iterator, batch_size)
+        return list(zip(*row_batch))
```

Comment:
Hah! Interview question used for real!

"How do you transpose a list of lists in python?"